crates/ruff_python_semantic/src/node.rs

```diff
@@ -0,0 +1,105 @@
+use std::num::TryFromIntError;
+use std::ops::{Index, IndexMut};
+
+use rustc_hash::FxHashMap;
+use rustpython_parser::ast::Stmt;
+
+use ruff_python_ast::types::RefEquality;
+
+/// Id uniquely identifying a statement in a program.
+///
+/// Using a `u32` is sufficient because Ruff only supports parsing documents with a size of max `u32::max`
+/// and it is impossible to have more statements than characters in the file.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct NodeId(u32);
+
+impl TryFrom<usize> for NodeId {
+    type Error = TryFromIntError;
+
+    fn try_from(value: usize) -> Result<Self, Self::Error> {
+        Ok(Self(u32::try_from(value)?))
+    }
+}
+
+impl From<NodeId> for usize {
+    fn from(value: NodeId) -> Self {
+        value.0 as usize
+    }
+}
+
+#[derive(Debug)]
+struct Node<'a> {
+    /// The statement this node represents.
+    stmt: &'a Stmt,
+    /// The ID of the parent of this node, if any.
+    parent: Option<NodeId>,
+    /// The depth of this node in the tree.
+    depth: u32,
+}
+
+/// The nodes of a program indexed by [`NodeId`]
+#[derive(Debug, Default)]
+pub struct Nodes<'a> {
```

Comment:
I wonder if we should rename `NodeId` to `StmtId` and `Nodes` to `Statements` since the API only accepts `Stmt`s as `node` parameters.

Reply:
I'm considering making this generic and using the same abstraction for expression-tracking.